pkg/virt-api/rest/streamer.go

```diff
@@ -0,0 +1,162 @@
+package rest
+
+import (
+	"context"
+	"io"
+	"net"
+	"time"
+
+	restful "github.com/emicklei/go-restful"
+	"github.com/gorilla/websocket"
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+)
+
+type vmiFetcher func(namespace, name string) (*v1.VirtualMachineInstance, *errors.StatusError)
+type dialer func(vmi *v1.VirtualMachineInstance) (net.Conn, *errors.StatusError)
+type validator func(vmi *v1.VirtualMachineInstance) *errors.StatusError
+type streamFunc func(clientConn *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult)
+type streamFuncResult error
+
+type Streamer struct {
+	fetchVMI        vmiFetcher
+	validateVMI     validator
+	dial            dialer
+	keepAliveClient func(ctx context.Context, conn *websocket.Conn, cancel func())
+
+	streamToClient streamFunc
+	streamToServer streamFunc
+}
+
+func NewRawStreamer(fetch vmiFetcher, validate validator, dial dialer) *Streamer {
+	return &Streamer{
+		fetchVMI:    fetch,
+		validateVMI: validate,
+		dial:        dial,
+		streamToServer: func(clientConn *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
+			_, err := io.Copy(serverConn, clientConn.UnderlyingConn())
+			result <- err
+		},
+		streamToClient: func(clientConn *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
+			_, err := io.Copy(clientConn.UnderlyingConn(), serverConn)
+			result <- err
+		},
+	}
+}
+
+func NewWebsocketStreamer(fetch vmiFetcher, validate validator, dial dialer) *Streamer {
+	return &Streamer{
+		fetchVMI:    fetch,
+		validateVMI: validate,
+		dial:        dial,
+		streamToServer: func(clientConn *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
+			_, err := kubecli.CopyFrom(serverConn, clientConn)
+			result <- err
+		},
+		streamToClient: func(clientConn *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
+			_, err := kubecli.CopyTo(clientConn, serverConn)
+			result <- err
+		},
+	}
+}
+
+func (s *Streamer) Handle(request *restful.Request, response *restful.Response) error {
+	namespace := request.PathParameter(NamespaceParamName)
+	name := request.PathParameter(NameParamName)
+
+	vmi, statusErr := s.fetchAndValidateVMI(namespace, name)
+	if statusErr != nil {
+		writeError(statusErr, response)
+		return statusErr
+	}
+
+	serverConn, statusErr := s.dial(vmi)
+	if statusErr != nil {
+		writeError(statusErr, response)
+		return statusErr
+	}
+	clientConn, err := clientConnectionUpgrade(request, response)
+	if err != nil {
+		writeError(errors.NewBadRequest(err.Error()), response)
```

Comment:
A thing I'm really not happy about is that we kind of abuse websockets and therefore have other problems.
As an example, from this point on we can no longer return errors over http to the client as the request is hijacked.

For SSH we might be able to still use the websocket control mechanism to return the errors, I haven't looked into that. But for anything doing raw streaming like VNC and Console errors happening after this point just collapse the streams.

Reply:
> A thing I'm really not happy about is that we kind of abuse websockets and therefore have other problems.
> As an example, from this point on we can no longer return errors over http to the client as the request is hijacked.

I am not sure I fully understand what you mean. We may not send too nice error messages at the moment and may 
 hide too much behind generic readers and writers, but at the end we send and receive usual websocket messages (in this case binary and close message). It should be mostly a matter of re-arranging code to send nice happy and unhappy close messages with the websocket from e.g. virt-handler. Once the websocket connection is established, there is no need to send something back via http. Either the connection is still healthy enough to send something back to the client via the websocket connection at this point or it is too broken anyway.

I see cases where it would be nice if we send something nicer back from `virt-api` if the connection to e.g. virt-handler breaks, but on the other hand that is not a situation which a client normally can interpret, since it is internal. Logging is most important and a clean close flow where we tell the client via the websocket close message that something unexpected happened, to avoid hanging messages or too abrupt tcp level closes.

Reply:
What I mean is that for vnc and console we proxy the websockets on a tcp level in virt-api which does not give us any control signals from the websockets themselves. That limits our options to detect problems in virt-api (like the keep-alive I added for SSH).
It's something virt-handler and the client have to do, but it's not as safe as doing the additional websocket hop in virt-api for vnc/console as well (I think).

Reply:
If there is something which you want to send (not sure what exactly) you can replace the current `io.Copy` with `UnderlyingConn()` with the read/writer in `staging/src/kubevirt.io/client-go/kubecli/websocket.go`.

Just add a function like this to `websocket.go`:

```go
func CopyOnWebsocketLayer(dst *websocket.Conn, src *websocket.Conn) (written int64, err error) {
	return io.Copy(&binaryWriter{conn: dst}, &binaryWriter{conn: src})
}
```